Improve scheduling, persist node details and add some tests

w-miller opened the pull request.

Commit:
Add new tenks_state_update plugin

Instead of just an allocations file, Tenks now uses a 'state' file. This
contains the node allocations, as well as physical network mappings and
indices. Clusters can also now be re-deployed with different specs to
update the state.

Commit:
Add inventory_hostname parameter to filter plugins

It is necessary in some places to use the network name filter plugins in
the context of any given hypervisor, even when the task is executing on
localhost. For this, allow an inventory_hostname to be specified.

Commit:
Persist physnet indices in state file

Instead of overwriting the physnet mappings specified in the Tenks
configuration, create a separate dict that maps physnet names to their
indices. Like physnet_mappings, this will be present for each
hypervisor.

Commit:
Update documentation

* Improve comments in tenks_update_state
* Add limitations section to README
* Add hosts blurb to README
* Add reconfiguration blurb to README

Commit:
Add basis for action plugin tests

Commit:
Fix mock import for Python 2

Commit:
Save task arguments and vars in object

Rather than passing these args through various sub-routines, make them
accessible through the action plugin object.

Commit:
Add coverage testing

Commit:
Add _process_specs tests

Commit:
Run coverage testing by default in tox

Commit:
Add additional _process_specs tests

Commit:
Remove unnecessary test

markgoddard commented on README.md:
```diff
@@ -13,6 +13,22 @@ installed inside it, Tenks' role dependencies can be installed by
+  `localhost`.
+* The `hypervisors` group should not directly contain any hosts. Its sub-groups
+  must contain one or more system. Systems in its sub-groups will host a subset
+  of the noides deployed by Tenks.
```
nodes

w-miller replied:
Thanks.

markgoddard commented on tests/test_tenks_update_state.py:
```diff
@@ -0,0 +1,245 @@
+        # After yet another run, the 'absent' state nodes should be deleted
+        # from state altogether.
+        self.mod._process_specs()
+        self.assertEqual(self.args['state']['foo']['nodes'], [])
```
Does this imply that scheduling twice in a row will remove the node from the state? What if the playbook fails?

w-miller replied:
Yes, unfortunately it does. I couldn't really think of a better way to do this, have you got any ideas?

w-miller replied:
I suppose we could leave absent nodes in the inventory forever but it might get a bit crufty...

markgoddard replied:
You could perform a second scheduling action at the end of the deploy.yml playbook that removes the nodes. Could modify the action to only remove absent nodes if a (new) flag is set.

w-miller replied:
That's true - I suppose that wouldn't even need to go through the Python action plugin if we just read from the file, used a Jinja `select` to remove absent nodes and wrote back to the file.

Commit:
Move absent node pruning to cleanup section

Add a cleanup_state.yml playbook that is run after every deployment and
teardown. It will remove any nodes marked as 'absent' from the state
file, by calling schedule.py with `prune_only=True`.

Commit:
Fix typo in README

Commit:
Add additional tox environments to Travis

Commit:
Add stestr config file

w-miller commented:
Travis checks finally working again - updated and ready for re-review.
